docs/proposals/2170-kubeflow-training-v2/README.md

```diff
@@ -0,0 +1,1362 @@
+
+// If the Elastic Policy is set, the numNodes parameter is ignored.
+// --nnodes=minNodes:maxNodes
+type TorchElasticPolicy struct {
+
+    // The limits to restart TrainJob.
+    // Insert it to the JobSet.spec.failurePolicy.maxRestarts
+    MaxRestarts *in32 `json:"maxRestarts,omitempty"`
+
+    // Min number of nodes for HPA and torchrun.
+    MinNodes *in32 `json:"minNodes,omitempty"`
+
+    // Max number of nodes for HPA and torchrun.
+    MaxNodes *in32 `json:"maxNodes,omitempty"`
+
+    // Metrics for scale up and down replicas.
+    Metrics []autoscalingv2.MetricSpec `json:"metrics,omitempty"`
+}
+```
+
+### The MPI Spec API
+
+The `MPISpec` API represents the configuration for training using MPI orchestration.
+E.g. creation of host-files and SSH keys. Using MPI might be more efficient for training on HPC
+clusters or for some ML frameworks (e.g. [MLX distributed with MPI](https://ml-explore.github.io/mlx/build/html/usage/distributed.html)).
+
+We will fully migrate to the MPI Operator V2 functionality as part of this KEP.
+Check [the proposal for the MPI V2 APIs.](https://github.com/kubeflow/mpi-operator/blob/master/proposals/scalable-robust-operator.md)
+
+```golang
+type MPISpec struct {
+    // Number of Procs per Node.
+    NumProcPerNode int `json:"numProcPerNode,omitempty"`
+
+    // MPI Implementation to create appropriate host-files.
+    // Can be one of OpenMPI, Intel, or MPICH.
+    MPIImplementation MPIImplementation `json:"mpiImplementation,omitempty"`
+
+    // Directory where SSH keys are mounted.
+    SSHAuthMountPath string `json:"SSHAuthMountPath,omitempty"`
+}
+
+type MPIImplementation string
+
+const (
+    MPIImplementationOpenMPI MPIImplementation = "OpenMPI"
+    MPIImplementationIntel   MPIImplementation = "Intel"
+    MPIImplementationMPICH   MPIImplementation = "MPICH"
+)
+```
+
+### Supported Runtimes by Community
+
+Kubeflow community are planning to support the following runtimes.
+
+#### PyTorch Distributed Runtime
+
+Initially, we will maintain only multi-node multi-worker runtime and PyTorch Elastic.
+
+```yaml
+apiVersion: kubeflow.org/v2alpha1
+kind: ClusterTrainingRuntime
+metadata:
+  name: torch-distributed-multi-node
+spec:
+  mlSpec:
+    torch:
+      numProcPerNode: 5
+  replicatedJobs:
+    - name: node
+      template:
+        spec:
+          template:
+            spec:
+              containers:
+                - name: trainer
+                  image: docker.io/kubeflow/pytorch-mnist
+                  env:
+                    - name: MASTER_ADDR
+                      value: "pytorch-node-0-0.pytorch"
+                    - name: MASTER_PORT
+                      value: 29400
+                  command:
+                    - torchrun train.py
+```
+
+Example of usage:
+
+```yaml
+apiVersion: kubeflow.org/v2alpha1
+kind: TrainJob
+metadata:
+  name: torch-distributed-multi-node
+  namespace: tenant-alpha
+spec:
+  trainingRuntimeRef:
+    name: pytorch-distributed
+  trainerConfig:
```

Comment:
This would be passed to the scheduler (eventually) - I'm a bit tired so the path there isn't totally clear - I'm guessing it will eventually be turned into units of pods with resource requests.